src/main/java/io/vertx/core/net/impl/TCPServerBase.java

```diff
@@ -153,14 +153,16 @@ public void updateTrafficShapingOptions(TrafficShapingOptions options) {
     if (options == null) {
       throw new IllegalArgumentException("Invalid null value passed for traffic shaping options update");
     }
-    if (trafficShapingHandler == null) {
-      throw new IllegalStateException("Unable to update traffic shaping options because the server was not configured " +
-                                      "to use traffic shaping during startup");
-    }
     TCPServerBase server = actualServer;
+    // Update the traffic shaping options only for the actual/main server
     if (server != null && server != this) {
-      server.updateTrafficShapingOptions(options);
+       server.updateTrafficShapingOptions(options);
```

Comment:
Can you update this, with `return` here to avoid repeated calls

Reply:
That is an interesting question. The repeated calls would happen if the consumer of this API invokes this method for all the shared servers. 

The decision here is based on the contract/interface we would like to expose to clients invoking `updateTrafficShapingOptions` (and similar APIs that affect configuration across shared servers)

1. Is it fair to expect clients to apply/change this setting for all the servers, while having the underlying implementation only update it when the actual server is involved? (which is what I think you are suggesting)

2. Alternatively, should we define the contract such that this setting should be applied/changed to ANY of the servers, and under the hood, vertx will only update it for the actual/main server?  (current implementation)

It seems like we currently have a model like (1). My only concern is that for invocations that are not on a actual/main server, it will do nothing (silently). 
If this is indeed the pattern we would like to follow, then I would prefer to have some signal (even if it were a log line) indicating that this was a worker server, so no updates were made. This might be useful, in the event that someone updates a server’s setting (that is not a main server) to not see the change taking effect.

 I see that `updateSSLOptions` does something similar to (1), although if it were called for all the servers, we would update the SSL options for the actual server N times (where N = no. verticles)

(2) comes with the assumption that this is a shared/global setting that applies to all servers, and hence can be updated on ANY of the servers. Although, one could argue that it is ambiguous for the client as to which server's configuration should be updated, and if they were to apply this to all servers, it would result in several redundant invocations for the actual server.
We could address this concern with redundant invocations by comparing the `TrafficShapingOptions` instances to only update when they differ. I would prefer to do something like this to have no ambiguity.

@vietj Do you have any thoughts or comments on this?

Reply:
In the latest commit, I made a change to only update the traffic-shaping options when they change. So, we address both concerns by:
1) Updating the actual-server's options regardless of which server is used for the update
2) Only perform the above update when the new options differ from the existing ones